### PR TITLE
Node events recording fixes

### PR DESCRIPTION
**Pull request by simon3z**

### CODE REVIEW

**Comment by simon3z** [comment]
@bgrant0607 I know you mentioned that nodes should not have a namespace (nor their events) but at this time pushing events to `/api/v1beta3/events` results in a 405 error.

**Comment by nikhiljindal** [comment]
As per http://kubernetes.io/third_party/swagger-ui/, namespace is not required while POSTing to /nodes (api/v1beta3/nodes), but it is required for POSTing to /events (/api/v1beta3/namespaces/default/events)

**Comment by bgrant0607** [comment]
@nikhiljindal is correct.

We haven't had time to fix the event namespace for nodes and don't have agreement on whether nodes should have a namespace.

**Comment by simon3z** [comment]
@bgrant0607 any idea about the plan? is this PR acceptable? would you prefer to enabling `/api/v1beta3/events` for node events?

**Comment by davidopp** [comment]
I'm sure I'm being dense, but why did you get rid of the other two events here?

**Comment by gmarek** [comment]
I was on vacation for 2 weeks and things changed so I may be missing some context, but I don't like the 'offline' word in this context. 'Unknown' is used in an API to denote third value and I'd rather be consistent with it. (This is not a strong opinion, so feel free to override me on this one)

I believe current condition is wrong and should be something similar to
```
if readyCondition.Status != lastReadyCondition.Status
```
because otherwise we loose information on moving from not-ready (e.g. Docker on Node has some problems or sth) to unknown. It should be good enough for now.

@davidopp IIUC removal of other two events here is justified by the way current logic of 'advancing' Node states is distributed - the only transition done by Node controller is to 'uknown' condition.

**Comment by simon3z** [comment]
@davidopp the reasons were explained in the commit messages:

> **nodecontroller: remove unused ready event recording**
> It's not nodecontroller setting the node to Ready so it's impossible to reach this condition.

> **nodecontroller: improve offline event recording**
> This patch substitutes the misleading reason "unknown" for the event recording. For symmetry with kubelet's message "online" the conditions Unknown and False are reported as "offline".

We could prefer `ready` and `notready` but actually having to comply with the event semantic "reason" I prefer `online` and `offline`.

**Comment by derekwaynecarr** [comment]
I am good with this fix and sticking events in the default namespace.

**Comment by bgrant0607** [comment]
I'm fine with using the default namespace.

"Offline" seems like an overly general characterization for "not ready".

**Comment by gmarek** [comment]
Are we intentionally avoid generating events for ```notready``` to ```unknown``` transition?

**Comment by davidopp** [comment]
I agree with Marek, this should be called "NodeReady==unknown" to be consistent with what users will see when they do "kubectl get nodes". As for Kubelet,
(1) the current recordNodeOnlineEvent() should really write "NodeReady==true"
(2) we should add recordNodeOfflineEvent() that should write "NodeReady==false" (right now it appears that Kubelet does not log the transition to NodeReady==false, only to NodeReady==true)

As for namespace, whatever works is fine with me as long as it doesn't make the normal use case harder (which is to pull all events using "kubectl get events")

**Comment by simon3z** [comment]
> I agree with Marek, this should be called "NodeReady==unknown" to be consistent with what users will see when they do "kubectl get nodes".

@davidopp what users see when they call `get nodes` is only `Ready` and `NotReady`, from `resource_printer.go`:

   if condition.Status == api.ConditionTrue {
           status = append(status, string(condition.Type))
   } else {
           status = append(status, "Not"+string(condition.Type))
   }

which are not events but a statuses.

When you transition to ConditionTrue the event should be `online` and when you transition to != ConditionTrue should be `offline` (these names recall better an event reason rather than a current status).

"NodeReady==unknown" is not seen by the user anywhere and it doesn't sound right for an event reason.

> (2) we should add recordNodeOfflineEvent() that should write "NodeReady==false" (right now it appears that Kubelet does not log the transition to NodeReady==false, only to NodeReady==true)

I didn't spot kubelet setting NodeReady==false, any pointer?

**Comment by gmarek** [comment]
That's only true for ordinary get call. If user asks for -o json, then she'll see 'real' status (I hope).

As for NodeReady == false -> not yet there, but is a v1.0 issue: #6285

**Comment by saad-ali** [comment]
This change (the namespace setting part) will not be needed once #7408 goes in.

**Comment by wojtek-t** [comment]
@piosz - since @gmarek will be ooo for some time, can you please take a look into it?

**Comment by piosz** [comment]
@saad-ali could you please confirm that the change is not needed and if yes close the bug?

**Comment by saad-ali** [comment]
@piosz The 2nd commit (`events: fix namespace events recording`) should be removed, it is not needed. I will defer to you on the nodecontroller changes.

**Comment by davidopp** [comment]
> I didn't spot kubelet setting NodeReady==false, any pointer?

That's weird, I could have sworn it did at the time I wrote that comment. Anyway, as @gmarek said we will add it for when Docker is dead (though I could have sworn it was previously set to false under some other condition(s)).

As for whether the event should be online/offline or online/offline/unknown, I still think it should be online/offline/unknown, but I didn't realize kubectl was printing Ready/NotReady, so I agree what you're doing here is more consistent with what we already have, so I'm fine with it. 

So, this change looks fine to me.

**Comment by simon3z** [comment]
Please let me know if anyone of you still prefers something else other than `online` and `offline`. Maybe `status-ready` and `status-unknown` (`ready` and `unknown` don't seem descriptive enough).

Basically if we wanted to make this valid for any condition the name would have to be `condition-ready-true` and `condition-ready-unknown`, but I don't know if we want to go down that road.

Overall I like best `online` and `offline` (or maybe `available`, `unavailable`?) but I am open to any proposal that is descriptive enough.

**Comment by bgrant0607** [comment]
`ready` and `!ready` ?

**Comment by simon3z** [comment]
@bgrant0607 don't you think we should scope it? `condition_ready` `!condition_ready`

**Comment by bgrant0607** [comment]
A better model to follow for an event is:
```
s.config.Recorder.Eventf(pod, "failedScheduling", "Binding rejected: %v", err)
```
There's a fixed, unique one-word reason (though it should be FailedScheduling, like our API constants) for this kind/type/category of event, and a more informative message, which may contain occurrence-specific details.

Events contain a reference to the "involved object", but I'm not sure that's sufficiently convenient for filtering events.

How about NodeNotReady?

**Comment by davidopp** [comment]
What's wrong with online/offline?

**Comment by bgrant0607** [comment]
online/offline is very non-specific and ambiguous, and completely disconnected from API concepts. We're going to have a number of conditions and/or knobs that can render a node unusable to some degree and/or for some workloads.

More generally, we need to develop a convention for event reason strings (should they contain the Kind or not, CamelCase vs. camelCase, etc.).

**Comment by davidopp** [comment]
OK sure, my point is just that we should either use existing terminology or invent something new that's very simple. Online/offline is in the second category; I agree it has the drawbacks you mentioned. So my suggestion would be that we change both the event generation and the resource printer to use { NodeReady, NotNodeReady, UnknownNodeReady }

This is similar to what @bgrant0607 proposed except the sense comes first (rather than going in the middle of the condition) and we surface Unknown rather than mapping it to Not.

**Comment by bgrant0607** [comment]
SGTM

**Comment by simon3z** [comment]
@davidopp @bgrant0607 can you review/merge? Thanks!

File: pkg/kubelet/kubelet.go [hunk 1]
```diff
@@ -1918,7 +1907,7 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 		if node.Status.Conditions[i].Type == api.NodeReady {
 			newCondition.LastTransitionTime = node.Status.Conditions[i].LastTransitionTime
 			if node.Status.Conditions[i].Status != api.ConditionTrue {
-				kl.recordNodeOnlineEvent()
+				kl.recordNodeStatusEvent("NodeReady")
```
Comment (davidopp) [inline pkg/kubelet/kubelet.go]: I must be missing something, but I don't understand the original logic (your change doesn't change the logic, just the way the event is recorded, but I want to make sure the logic is right). From the previous block of code in this function, it seems that the newCondition may be NodeReady==True or NodeReady==False. Yet the logic here seems to say "if the old condition was NodeReady!=True, then record NodeReady" even if the new condition is also NodeReady!=True. It also seems to not be updating LastTransitionTime, which seems wrong in the case where the new condition != the old condition.

The next line you changed (in the !updated section) also seems to be wrong for similar reasons.

The logic used for Unschedulable seems correct and is what I would have expected for NodeReady.

Can you explain what I'm missing?

**Comment by simon3z** [comment]
@davidopp thanks for the review, the logic should be fine now.

File: pkg/kubelet/kubelet.go [hunk 2]
```diff
@@ -1916,25 +1906,28 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 	updated := false
 	for i := range node.Status.Conditions {
 		if node.Status.Conditions[i].Type == api.NodeReady {
-			newCondition.LastTransitionTime = node.Status.Conditions[i].LastTransitionTime
-			if node.Status.Conditions[i].Status != api.ConditionTrue {
-				kl.recordNodeOnlineEvent()
-			}
-			node.Status.Conditions[i] = newCondition
+			oldReadyNodeConditionStatus = node.Status.Conditions[i].Status
```
Comment (davidopp) [inline pkg/kubelet/kubelet.go]: I think this logic is still wrong. I think what we want is something like
```
if (old == new) {
   newReadyNodeCondition.LastTransitionTime = node.Status.Conditions[i].LastTransitionTime
}
else {
  node.Status.Conditions[i] = newReadyNodeCondition
  newReadyNodeCondition.LastTransitionTime = currentTime
}
```

Comment (simon3z) [inline pkg/kubelet/kubelet.go]: I think that:

    newReadyNodeCondition.LastUpdateTime = currentTime

is not needed as it's already set above (`LastHeartbeatTime`)

Also about: `// also generate appropriate event`, I prefer to keep the logic in just one place (below).

File: pkg/kubelet/kubelet.go [hunk 3]
```diff
@@ -1889,9 +1878,10 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 	}()
 
 	currentTime := util.Now()
-	var newCondition api.NodeCondition
+	var newReadyNodeCondition api.NodeCondition
```
Comment (davidopp) [inline pkg/kubelet/kubelet.go]: nits:
newReadyNodeCondition -> newNodeReadyCondition
oldReadyNodeConditionStatus -> oldNodeReadyConditionStatus

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -1916,25 +1906,28 @@ func (kl *Kubelet) setNodeStatus(node *api.Node) error {
 	updated := false
 	for i := range node.Status.Conditions {
 		if node.Status.Conditions[i].Type == api.NodeReady {
-			newCondition.LastTransitionTime = node.Status.Conditions[i].LastTransitionTime
-			if node.Status.Conditions[i].Status != api.ConditionTrue {
-				kl.recordNodeOnlineEvent()
-			}
-			node.Status.Conditions[i] = newCondition
+			oldReadyNodeConditionStatus = node.Status.Conditions[i].Status
+			newReadyNodeCondition.LastTransitionTime = node.Status.Conditions[i].LastTransitionTime
```
Comment (davidopp) [inline pkg/kubelet/kubelet.go]: This line is wrong. It should be
```
if (oldReadyNodeConditionStatus == newReadyNodeCondition.Status) {
   newReadyNodeCondition.LastTransitionTime = node.Status.Conditions[i].LastTransitionTime
} else {
  newReadyNodeCondition.LastTransitionTime = currentTime
}
```

**Comment by davidopp** [comment]
After making the changes I suggested, please squash the commits and add to the commit message "Fixes #8585" (though that's a bit deceptive since that wasn't the original purpose of this PR).

File: pkg/kubelet/kubelet.go (hunk 2 above)
Comment (davidopp) [inline pkg/kubelet/kubelet.go]: Yeah, never mind. I think the only remaining logic problem is the one I just commented on.

**Comment by simon3z** [comment]
@davidopp ok I'll make another PR on top of this for fixing #8585, can we merge this?

**Comment by davidopp** [comment]
I wasn't suggesting to fix #8585 in a separate PR; in fact exactly the opposite. Is it possible to re-apply that commit? It was completely correct except for the one issue I identified.

**Comment by simon3z** [comment]
@davidopp ok thanks, hopefully I've handled it all.

**Comment by simon3z** [comment]
Sorry @davidopp let's see if I got it straight this time :smile:

**Comment by davidopp** [comment]
LGTM

**Pull request merged**